### PR TITLE
Fix FileExist utility and generate sha for modelspec

### LINKED ISSUE #1203: MMS model agent gets panic when downloading models
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

```
{"level":"info","ts":1605215785.7493982,"logger":"Downloader","msg":"Downloading to model dir","modelUri":"s3://torchscript","modelDir":"/mnt/models"}
{"level":"info","ts":1605215785.7495825,"logger":"modelAgent","msg":"Download model ","modelName":"cifar10","storagUri":"s3://torchscript","modelDir":"/mnt/models"}
panic: runtime error: index out of range [1] with length 1

goroutine 22 [running]:
github.com/kubeflow/kfserving/pkg/agent/storage.(*S3Provider).DownloadModel(0xc00048d9a0, 0x7ffed6e0f493, 0xb, 0xc000580480, 0x7, 0xc0005804a0, 0x10, 0x1b, 0x0)
	github.com/kubeflow/kfserving/pkg/agent/storage/s3.go:59 +0x49c
github.com/kubeflow/kfserving/pkg/agent.(*Downloader).download(0xc000286198, 0xc000580480, 0x7, 0xc0005804a0, 0x10, 0x2, 0x3a)
	github.com/kubeflow/kfserving/pkg/agent/downloader.go:75 +0xff
github.com/kubeflow/kfserving/pkg/agent.(*Downloader).DownloadModel(0xc000286198, 0xc000580480, 0x7, 0xc00045c540, 0x0, 0x0)
	github.com/kubeflow/kfserving/pkg/agent/downloader.go:51 +0x76a
github.com/kubeflow/kfserving/pkg/agent.(*Puller).modelProcessor(0xc000286180, 0xc000580480, 0x7, 0xc00045c5a0)
	github.com/kubeflow/kfserving/pkg/agent/puller.go:131 +0x2b9
created by github.com/kubeflow/kfserving/pkg/agent.(*Puller).enqueueModelOp
```

**What did you expect to happen:**

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Istio Version:
- Knative Version:
- KFServing Version:
- Kubeflow version:
- Kfdef:[k8s_istio/istio_dex/gcp_basic_auth/gcp_iap/aws/aws_cognito/ibm]
- Minikube version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### LINKED ISSUE #1317: invalid memory address or nil pointer dereference in agent
/kind bug

**What steps did you take and what happened:**

We deployed the following triton inferenceservice:

```yaml
apiVersion: "serving.kubeflow.org/v1beta1"
kind: "InferenceService"
metadata:
  name: "triton-mms"
spec:
  predictor:
    triton:
      args:
      - --log-verbose=1
      resources:
        limits:
          cpu: "1"
          memory: "8Gi"
        requests:
          cpu: "1"
          memory: 8Gi
```

as well as the following TrainedModel

```yaml
apiVersion: "serving.kubeflow.org/v1alpha1"
kind: "TrainedModel"
metadata:
  name: "ourmodelname"
spec:
  inferenceService: triton-mms
  model:
    framework: pytorch
    storageUri: s3:/path-to-container/ourmodelname
    memory: "1Gi"
```

within this s3 container we have stored the following mode:

```
- ourmodelname/
  |_ config.pbtxt
  |_ /1
     |_ model.pt
```

we then observe the following error in the `agent` when loading the mode:
```
{"level":"info","ts":"2021-01-26T08:45:42.264Z","caller":"agent/puller.go:121","msg":"Worker is started for layout"}
{"level":"info","ts":"2021-01-26T08:45:42.264Z","caller":"agent/puller.go:129","msg":"Downloading model from s3:/path-to-container/ourmodelname"}
{"level":"info","ts":"2021-01-26T08:45:42.264Z","caller":"agent/downloader.go:47","msg":"Downloading s3:/path-to-container/ourmodelname to model dir /mnt/models"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x169f489]

goroutine 70 [running]:
github.com/kubeflow/kfserving/pkg/agent/storage.FileExists(0xc0001a4480, 0x7b, 0x1bfab30)
	/go/src/github.com/kubeflow/kfserving/pkg/agent/storage/utils.go:27 +0x99
github.com/kubeflow/kfserving/pkg/agent.(*Downloader).DownloadModel(0xc0003eaa98, 0xc0007c5288, 0x6, 0xc0001a90e0, 0x0, 0x0)
	/go/src/github.com/kubeflow/kfserving/pkg/agent/downloader.go:49 +0x40f
github.com/kubeflow/kfserving/pkg/agent.(*Puller).modelProcessor(0xc0003eaa80, 0xc0007c5288, 0x6, 0xc0001a95c0)
	/go/src/github.com/kubeflow/kfserving/pkg/agent/puller.go:130 +0x273
created by github.com/kubeflow/kfserving/pkg/agent.(*Puller).enqueueModelOp
	/go/src/github.com/kubeflow/kfserving/pkg/agent/puller.go:89 +0x120
```

Interestingly, when we try the same with the example model `cifar10` we do not observe this issue.

**What did you expect to happen:**
Model is loaded successfully 

**Environment:**

- Knative Version: 0.17
- KFServing Version: 0.5.0-rc2

### PR DESCRIPTION
**Pull request by yuzisun**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1317 

**Special notes for your reviewer**:

1. Generate sha for the model spec instead of individual fields
2. Fix FileExist util function as there could be other PathError if not `FileNotExist`

**Release note**:

```release-note

```

### CODE REVIEW

File: pkg/agent/downloader.go [hunk 1]
```diff
@@ -38,25 +39,25 @@ var SupportedProtocols = []storage.Protocol{storage.S3, storage.GCS}
 
 func (d *Downloader) DownloadModel(modelName string, modelSpec *v1alpha1.ModelSpec) error {
 	if modelSpec != nil {
-		modelUri := modelSpec.StorageURI
-		hashModelUri := hash(modelUri)
-		hashFramework := hash(modelSpec.Framework)
-		hashMemory := hash(modelSpec.Memory.String())
+		sha256 := storage.AsSha256(modelSpec)
```
Comment (ifilonenko) [inline pkg/agent/downloader.go:42]: if we are changing hashing strategy the unhashing needs to be handled as well

**Review by ifilonenko — CHANGES_REQUESTED** [review]

**Comment by rmporsch** [comment]
Thanks so much @yuzisun for taking care of this!

**Pull request merged**